### PR TITLE
normcap: init at 0.4.4

### LINKED ISSUE #166228: Packaging request: normcap
**Project description**
OCR-powered screenshot tool to capture text instead of images. 

**Metadata**

* homepage URL: https://dynobo.github.io/normcap/
* source URL: https://github.com/dynobo/normcap
* license: GPL3
* platforms: unix
* example package: [AUR](https://aur.archlinux.org/packages/normcap) ([PKGBUILD](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=normcap))

### PR DESCRIPTION
**Pull request by cafkafk**
## Description of changes

Closes: #166228

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/graphics/normcap/default.nix [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  propagatedBuildInputs = with python3.pkgs; [
+    pyside6
+
+    # Test
+    toml
+    pytest-qt
+  ];
+
+  postPatch = ''
+    sed -i 's|PySide6-Essentials = "6.5.1"||' pyproject.toml
```
Comment (TheBrainScrambler) [inline pkgs/applications/graphics/normcap/default.nix:34]: Use `substituteInPlace` instead.
https://nixos.org/manual/nixpkgs/stable/#fun-substitute

**Review by TheBrainScrambler — CHANGES_REQUESTED** [review]
Not sure if anything else is correct, but I saw that low-hanging fruit

**Comment by TheBrainScrambler** [comment]
Also why is this a draft ? Does the package fail to build ?

**Comment by cafkafk** [comment]
> Also why is this a draft ? Does the package fail to build ?

It seems to build on ofborg, but I couldn't test locally because cache.nixos.org doesn't seem to have a dependency I needed and building it took an unreasonable amount of time, so I was waiting for it to be cached.

If anyone else can test this locally and report back if it's working I'd mark it ready for review.

**Comment by TheBrainScrambler** [comment]
Ah yes shiboken6, I saw that an update for it merged recently. I'm building it right now.

File: pkgs/applications/graphics/normcap/default.nix [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  '';
+
+  meta = with lib; {
+    description = "OCR powered screen-capture tool to capture information instead of images";
+    homepage = "https://dynobo.github.io/normcap/";
+    license = licenses.gpl3;
```
Comment (TheBrainScrambler) [inline pkgs/applications/graphics/normcap/default.nix:40]: `gpl3` is deprecated. According to https://github.com/dynobo/normcap/blob/main/LICENSE this should be `gpl3Plus`

File: pkgs/applications/graphics/normcap/default.nix [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    maintainers = with maintainers; [cafkafk];
```
Comment (TheBrainScrambler) [inline pkgs/applications/graphics/normcap/default.nix:41]: `[cafkafk]` -> `[ cafkafk ]`

**Review by TheBrainScrambler — CHANGES_REQUESTED** [review]
As usual I got a bit fast on my review

**Comment by TheBrainScrambler** [comment]
Well then there was pyside6, and I confirm that this one takes too long ... So I will do like you did, not building it and waiting for it to be in cache.

**Comment by TheBrainScrambler** [comment]
Result of `nixpkgs-review pr 253578` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>normcap</li>
    <li>normcap.dist</li>
  </ul>
</details>

**Comment by TheBrainScrambler** [comment]
pyside6 is finally cached so it's possible to build this now. For some reasons i also get a `normcap.dist` package for free ?

When running normcap though it fails with this error:

```
15:20:57 - CRITICAL - normcap:161 - Uncaught exception! Quitting NormCap!
15:20:57 - CRITICAL - normcap:187 - Uncaught exception! Quitting NormCap! (debug output limited)
Traceback (most recent call last):
  File "/nix/store/8w24h7wh234d4ijwx4yki41ppz17wc6d-normcap-0.4.4/bin/.normcap-wrapped", line 9, in <module>
    sys.exit(main())
  File "/nix/store/8w24h7wh234d4ijwx4yki41ppz17wc6d-normcap-0.4.4/lib/python3.10/site-packages/normcap/app.py", line 57, in main
    tray = SystemTray(app, vars(args))
  File "/nix/store/8w24h7wh234d4ijwx4yki41ppz17wc6d-normcap-0.4.4/lib/python3.10/site-packages/normcap/gui/tray.py", line 60, in __init__
    logger.debug("System info:\n%s", system_info.to_dict())
  File "/nix/store/8w24h7wh234d4ijwx4yki41ppz17wc6d-normcap-0.4.4/lib/python3.10/site-packages/normcap/gui/system_info.py", line 165, in to_dict
    "tesseract_path": get_tesseract_path(),
  File "/nix/store/8w24h7wh234d4ijwx4yki41ppz17wc6d-normcap-0.4.4/lib/python3.10/site-packages/normcap/gui/system_info.py", line 81, in get_tesseract_path
    raise RuntimeError(
RuntimeError: No Tesseract binary found! Tesseract has to be installed and added to PATH environment variable.
15:20:57 - CRITICAL - normcap:192 - Please open an issue with the output above on https://github.com/dynobo/normcap/issues
```

Obviously it's missing the `tesseract` dependency, so you will have to add that.
To avoid more problems like that have a look at the AUR package at https://aur.archlinux.org/packages/normcap and its dependencies, this might help.

**Comment by cafkafk** [comment]
I added `tesseract4`, it seems to work now (added `leptonica` as well, but it seemed to work without it). Also added `wl-clipboard` for Wayland users, but I can't test that myself. This does work now, but on my very janky XMonad it can be a bit weird, and I'll likely have to add some fixes locally to make it neat.

Ideally someone would test this on:
- [ ] Wayland, does the clipboard work?
- [ ] KDE, Gnome, or mainstream DE.

> For some reasons i also get a normcap.dist package for free ?

I've got no idea why tbh, hopefully some reviewer can look into this.

File: pkgs/applications/graphics/normcap/default.nix (hunk 1 above)
Comment (cafkafk) [reply pkgs/applications/graphics/normcap/default.nix]: I tried in the last push, can you check that it's correct?

Comment (TheBrainScrambler) [reply pkgs/applications/graphics/normcap/default.nix]: Yes, perfect !

**Comment by TheBrainScrambler** [comment]
I've tried again `nixpkgs-review` with it and it still fails with this error saying tesseract wasn't found. Looking at your nix file it's because you put tesseract4 in `nativeBuildInputs`. Packages in `nativeBuildInputs` are only available during build time. You have to move it (and leptonica too I guess) to `propagatedBuildInputs`.

I suspect it worked for you because you used a nix-shell ? Next time try it with nix-build to be sure

**Comment by cafkafk** [comment]
> I've tried again `nixpkgs-review` with it and it still fails with this error saying tesseract wasn't found. Looking at your nix file it's because you put tesseract4 in `nativeBuildInputs`. Packages in `nativeBuildInputs` are only available during build time. You have to move it (and leptonica too I guess) to `propagatedBuildInputs`.
> 
> I suspect it worked for you because you used a nix-shell ? Next time try it with nix-build to be sure. You can also use nixpkgs-review.

I updated it. Ran `nix build ./#normcap` and ran the binary `./result/bin/normcap`, and it worked. I'm not using a nix-shell thou, so I do wonder why it worked, before I tested via `nix profile install ./#normcap`.

**Comment by TheBrainScrambler** [comment]
Result of `nixpkgs-review pr 253578` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>normcap</li>
    <li>normcap.dist</li>
  </ul>
</details>

**Comment by TheBrainScrambler** [comment]
Strange, maybe you have tesseract installed and it leaked in ? I don't know.
In any case it works for me too now, so it should be good.

**Review by TheBrainScrambler — APPROVED** [review]

File: pkgs/applications/graphics/normcap/default.nix [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+  src = fetchFromGitHub {
+    owner = "dynobo";
+    repo = "normcap";
+    rev = "v${version}";
+    hash = "sha256-dShtmoqS9TC3PHuwq24OEOhYfBHGhDCma8Du8QCkFuI=";
+  };
+
+  buildInputs = [
+    wl-clipboard
```
Comment (pbsds) [inline pkgs/applications/graphics/normcap/default.nix:25]: Does this include the x11 fallback tools?

Comment (cafkafk) [reply pkgs/applications/graphics/normcap/default.nix:25]: ¯\\_(ツ)_/¯ I have no idea tbh

**Comment by tennox** [comment]
```
$ normcap
16:53:05 - ERROR   - normcap:211 - [QT] qtfatalmsg - this application failed to start because no qt platform plugin could be initialized. reinstalling the application may fix this problem.

available platform plugins are: eglfs, linuxfb, minimal, minimalegl, offscreen, vkkhrdisplay, vnc, xcb.

fish: Job 1, 'normcap' terminated by signal SIGABRT (Abort)
```

Am I doing something wrong?
Or is this an issue?

**Comment by TheBrainScrambler** [comment]
I have no idea, for me it works ...
Can't be a regression though since there hasn't been any update for this
package.

**Comment by cafkafk** [comment]
> ```
> $ normcap
> 16:53:05 - ERROR   - normcap:211 - [QT] qtfatalmsg - this application failed to start because no qt platform plugin could be initialized. reinstalling the application may fix this problem.
> 
> available platform plugins are: eglfs, linuxfb, minimal, minimalegl, offscreen, vkkhrdisplay, vnc, xcb.
> 
> fish: Job 1, 'normcap' terminated by signal SIGABRT (Abort)
> ```
> 
> Am I doing something wrong? Or is this a regression?

Might wanna create a issue, I can't reproduce this.

**Comment by yorickvP** [comment]
wayland support, I bet, try `QT_QPA_PLATFORM=xcb normcap`

**Comment by tennox** [comment]
> wayland support, I bet, try `QT_QPA_PLATFORM=xcb normcap`

Thanks, @yorickvP - that fixes it.
Is there a way to bake this into the nixpkgs package?
Found many different ways this env var is set, and tried to find those that are also a python package, but the python ones only set it to e.g. `offscreen` in the check/install phase:
- [vorta](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/backup/vorta/default.nix#L71) (setting in preCheck)
Other apps use e.g.:
- [`qtWrapperArgs`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/video/plex-media-player/default.nix#L42)
- [`wrapProgram`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/networking/instant-messengers/viber/default.nix#L82)

I guess the latter might be most promising, but I'm fishing in unknown waters here :thinking:

**Pull request merged**